### ISSUE TITLE
When TLS transport is in writeBlockedOnRead=True mode, data is buffered in memory but producer APIs are not respected

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5063]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5063|
|Type|release blocker: regression|
|Created|2011-04-22 22:57:48Z|
|Branch|https://github.com/twisted/twisted/tree/tls-producer-5063-2|

If _writeBlockedOnRead=True, writes to the TLS transport are buffered in memory... but if a producer is registered it will never get pauseProducing called. This means data might be accumulated in memory without application code ever knowing it should stop writing.

<details><summary>Searchable metadata</summary>

```
trac-id__5063 5063
type__release_blocker__regression release blocker: regression
reporter__itamar itamar
priority__high high
milestone__Twisted_11_1 Twisted-11.1
branch__branches_tls_producer_5063_2 branches/tls-producer-5063-2
branch_author__itamar__exarkun itamar, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1303513068000000 1303513068000000
changetime__1311596604000000 1311596604000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [31942]) Branching to 'tls-producer-5063'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

This branch, which also covers [#5093](https://github.com/twisted/twisted/issues/5093), is basically done. Before being put up for review it just needs to be merged forward once [#5118](https://github.com/twisted/twisted/issues/5118) is fixed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [32339]) Branching to 'tls-producer-5063-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Ready for review. Note that this branch also fixes [#5093](https://github.com/twisted/twisted/issues/5093), so the review should address the issues covered by both tickets, namely: producers were broken in writeBlockedOnRead, were broken when not in writeBlockedOnRead mode, and had no tests.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

I just realized there's another edge case I didn't handle...

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

And, done. Test results:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tls-producer-5063-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. `twisted/protocols/tls.py`
    1. `_PullToPush.startStreaming` has a stray import
    1. The `log.err` calls (all `log.err` calls, really) should have a message argument passed to them so if they show up in logs, their source can more easily be identified.
    1. The new `_streamingProducer` attribute of `TLSMemoryBIOProtocol` looks like it isn't needed
    1. The new `_producer` attribute is undocumented
    1. In `TLSMemoryBIOProtocol.write`, the way writes are ignored after disconnect seems surprising.  I suppose this isn't a huge deal, but it would be nice if it were the same check abstract.py does: `if not self.connected or self._writeDisconnected:` - which maybe just means `if not self.connected:` since I don't think we have TLS half close.  Of course, I guess we also don't have `connected` here either.
    1. If there were a mostly-passthrough implementation of the `_PullToPush` interface for `IPushProducer`, some of the producer support code in `TLSMemoryBIOProtocol` would be simpler.  It could just call `startStreaming` and `stopStreaming` unconditionally (this would particularly improve the `stopStreaming` case).
  1. Generally, test coverage looks pretty great.  But:
    1. It looks like `LogProtocol` in `buildTLSProtocol` doesn't need to be nested
    1. Also, the print in `LogProtocol.connectionLost` needs to be something else
    1. Also, I bet you can use `twisted.test.proto_helpers.AccumulatingProtocol` instead of `LogProtocol`.
    1. It's super non-obvious why `test_streamingProducerPausedInWriteBlockedOnReadMode` triggers a `WantReadError`, might merit some more commentery.
    1. `test_streamingProducerResumedInWriteBlockedOnReadMode` works without the loop around `self.flushTwoTLSProtocols`
    1. Typo in `loseConnectionWithProducer` - "Undelying"
    1. `test_nonStreamingProducer` could use `assertIsInstance`
    1. `test_nonStreamingProducer` also appears not to need the range(1000) loop
    1. `flushTwoTLSProtocols` likely only ever needs 3 iterations to complete the handshake.
    1. both `test_interface` would probably be better using `verifyObject` rather than `Interface.providedBy` (and the former implies the latter)
    1. `test_pause` could use `self.fail` in `shouldNotBeCalled` instead of `self.assertTrue(False)`
    1. in `test_resume`, `reactor.callLater(0.01, self.producer.resumeProducing)`: booooooooooooooooooooo fail

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

Re 1.5, I'm probably going to pass, since I don't want to implement a "connected" attribute just for that. If you want I can open a ticket for the long awaited "tell me the connection state of this transport damnit" API.

In my local checkout I implemented _PushToPush, a transparent wrapper that does stop/startStreaming... and that plus tests is much more code than just having an "if isinstance" in unregisterProducer. More code for such a tiny benefit doesn't seem worth it, so I'm inclined to reject 1.6.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst removed owner|
|-|-|

All review comments have been addressed, except for 1.5 and 1.6 (see explanations in previous comment). I think "fixing" 2.12 made the test not as good, but it still does fail if resumeProducing is commented out, so I can live with it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @itamarst|
|-|-|

I understand and agree regarding 1.5.  For 1.6, I don't mind if the code goes in as-is, but I think that even if `_PushToPush` is more lines of code it's still better, since it removes one of the possible branches.  For 2.12, real time based tests always suck, so I'm glad it does what it does now. :)

So, looks good, please merge.  Thanks for working on this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst set status to `closed`|
|-|-|

(In [32362]) Merge tls-producer-5063-2.

Fixes: [#5063](https://github.com/twisted/twisted/issues/5063), [#5093](https://github.com/twisted/twisted/issues/5093)
Author: itamar, exarkun
Reviewer: exarkun

twisted.protocols.tls (and therefore new reactor TLS codepath) now supports producers correctly.
